File: ml-benchmark/synduce/list/slsp.f
Inductive List = elt Int | cons {Int, List};
Inductive CnList = sglt Int | cat {CnList, Int, CnList};

max = \x: Int. \y: Int. if (> x y) then x else y;

cat_list = fix (
  \f: List -> List -> List. \xs: List. \ys: List.
  match xs with
    elt a -> cons {a, ys}
  | cons {hd, tl} -> cons {hd, f tl ys}
  end
);

repr = fix (
  \f: CnList -> Compress List. \xs: CnList.
  match xs with
    sglt a -> elt a
  | cat {a, b, c} -> cat_list (f a) (f c)
  end
);

spec = \xs: List. (fix (
  \f: List -> {Int, Bool}. \xs: List.
  match xs with
    elt a -> {max 0 a, (>= a 0)}
  | cons {hd, tl} ->
    let result = f tl in
    let new_cond = and result.2 (>= hd 0) in
    {if new_cond then + hd result.1 else result.1, new_cond}
  end
) xs).1;

main = \x: CnList. spec (repr x);

File: ml-benchmark/synduce/list/slsp.ml
type tlist = Elt of int | Cons of int * tlist
type cnlist = Sglt of int | Cat of cnlist * int * cnlist

let max x y = if x > y then x else y

let rec cat_list xs ys =
  match xs with
  | Elt a -> Cons (a, ys)
  | Cons (hd, tl) -> Cons (hd, cat_list tl ys)

let rec repr xs =
  match xs with
  | Sglt a -> Elt a
  | Cat (a, b, c) -> cat_list (repr a) (repr c)

let spec xs =
  let rec f xs =
    match xs with
    | Elt a -> (max 0 a, a >= 0)
    | Cons (hd, tl) ->
        let result = f tl in
        match result with
        | (s, b) ->
            let new_cond = b && hd >= 0 in
            if new_cond then (hd + s, new_cond) else (s, new_cond)
  in
  match f xs with
  | (s, b) -> s

let program x = spec (repr x)